File: kontra/snake.js
kontra.init();
            
// variables
SPRITE_SIZE = 20
SCREEN_SIZE = 20;
INITIAL_TAIL = 5;
FPS = 15

let apple = kontra.sprite({
    x: 3,
    y: 3,

    // required for a rectangle sprite
    width: SPRITE_SIZE,
    height: SPRITE_SIZE,
    color: 'red',
    
    render: function() {
        this.context.fillStyle = this.color;
        this.context.fillRect(this.x * SPRITE_SIZE + 1, this.y * SPRITE_SIZE + 1, this.width - 1, this.height - 1);
    }
});

let snake = kontra.sprite({
    x: 10,
    y: 10,           
    tail: INITIAL_TAIL,
    trail: [],

    // required for a rectangle sprite
    width: SPRITE_SIZE,
    height: SPRITE_SIZE,
    color: 'green',
    
    checkCollision: function() {
        this.trail.forEach(element => {
            if(element.x == this.x && element.y == this.y) {
                if(snake.tail > INITIAL_TAIL){
                    console.log("Scored: %i", this.tail - INITIAL_TAIL);
                }
                this.x = this.y = 10;
                this.dx = this.dy = 0;
                this.tail = INITIAL_TAIL;
            }
        });
    },

    update: function() {
        this.x = this.x > SCREEN_SIZE -1 ? 0 : this.x < 0 ? SCREEN_SIZE - 1 : this.x;
        this.y = this.y > SCREEN_SIZE -1 ? 0 : this.y < 0 ? SCREEN_SIZE - 1 : this.y;                

        this.trail.push({ x: this.x, y: this.y});
        this.advance();
        
        while(this.trail.length > this.tail){
            this.trail.shift();
        }

        this.checkCollision();
    },
    
    render: function() {
        this.context.fillStyle = this.color;
        this.trail.forEach(element => {
            this.context.fillRect(element.x * SPRITE_SIZE + 1, element.y * SPRITE_SIZE + 1, this.width - 1, this.height - 1);
        });
    }
});

function checkInput() {
    if(snake.dx == 0){
        if(kontra.keys.pressed('left')) {
            snake.dx = -1;
            snake.dy = 0;
        }
        if(kontra.keys.pressed('right')) {
            snake.dx = 1;
            snake.dy = 0;
        }
    }
    if(snake.dy == 0) {
        if(kontra.keys.pressed('down')) {
            snake.dx = 0;
            snake.dy = 1;
        }
        if(kontra.keys.pressed('up')) {
            snake.dx = 0;
            snake.dy = -1;
        }
    }
}

function generateApple() {
    needNewApple = false;
    do {
        needNewApple = false;
        apple.x = Math.floor(Math.random() * SCREEN_SIZE);
        apple.y = Math.floor(Math.random() * SCREEN_SIZE);
        snake.trail.forEach(element => {
            needNewApple |= (element.x == apple.x && element.y == apple.y);
        });
    } while(needNewApple)
}

function checkPickup() {
    if(apple.x == snake.x && apple.y == snake.y){
        snake.tail++;
        generateApple();
    }
}

let loop = kontra.gameLoop({
    fps: FPS,
    clearCanvas: true,

    update: function() {
        checkInput();
        checkPickup();
        snake.update();
    },

    render: function() {
        snake.render();
        apple.render();
    }
});

loop.start();

File: rust/sdl2/src/game_objects/snake.rs
use std::collections::VecDeque;
use sdl2::pixels::Color;
use sdl2::rect::{Rect};
use sdl2::render::{Canvas};
use sdl2::video::{Window};

use game_objects::position::{Position};
use game_objects::constants::{SPRITE_SIZE, SCREEN_SIZE, INITIAL_TAIL};

pub struct Snake {
    pub x: i32,
    pub y: i32,
    pub dx: i32,
    pub dy: i32,
    pub trail: VecDeque<Position>,
    pub tail: usize
}

impl Snake {

    pub fn check_collision(&mut self, x: i32, y: i32) -> bool {
        for element in &self.trail {
            if element.x == x && element.y == y {
                return true;
            }
        }
        return false;
    }
    
    pub fn update(&mut self) {
        let x = (self.x + SCREEN_SIZE + self.dx) % SCREEN_SIZE;
        let y = (self.y + SCREEN_SIZE + self.dy) % SCREEN_SIZE;
        if self.check_collision(x, y) {
            self.tail = INITIAL_TAIL;
            self.x = 10;
            self.y = 10;
            self.dx = 0;
            self.dy = 0;
        } else {
            self.x = x;
            self.y = y;
        }

        self.trail.push_back(Position::new(self.x, self.y));
        loop {
            if self.trail.len() <= self.tail {
                break;
            }
            self.trail.pop_front();
        }
    }

    pub fn draw(&mut self, canvas: &mut Canvas<Window>) {
        canvas.set_draw_color(Color::RGB(0,255,0));
        let width = (SPRITE_SIZE - 1) as u32;
        let height = (SPRITE_SIZE - 1) as u32;
        for element in &self.trail {
            canvas.fill_rect(
                Rect::new(element.x * SPRITE_SIZE + 1, element.y * SPRITE_SIZE + 1, width, height))
                .expect("Failed to draw the snake");
        }
    }
}
